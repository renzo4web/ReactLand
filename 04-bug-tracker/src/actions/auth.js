import Swal from 'sweetalert2';

import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';
import { bugsLogout } from './bugs';
import { finishLoading, startLoading } from './ui';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());

        firebase
            .auth()
            .signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());
            })
            .catch((err) => {
                dispatch(finishLoading());
                Swal.fire('Error', err.message);
            });
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase
            .auth()
            .signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                const { displayName, uid } = user;
                dispatch(login(uid, displayName));
            });
    };
};

export const startRegisterWithEmailPasswordName = (name, email, password) => {
    return (dispatch) => {
        firebase
            .auth()
            .createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({
                    displayName: name,
                });

                dispatch(login(user.id, user.displayName));
            })
            .catch((err) => {
                Swal.fire('Error', err.message);
            });
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName,
    },
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(bugsLogout());
        dispatch(logout());
    };
};

export const logout = () => ({
    type: types.logout,
});
