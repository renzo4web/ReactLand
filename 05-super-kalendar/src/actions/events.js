import toast from 'react-hot-toast';
import { fetchWithToken } from '../helpers/fetch';
import { prepareEvents } from '../helpers/prepareEvents';
import { types } from '../types/types';

/*
    ADDING THE EVENTS TO DB
*/

export const eventStartAddNew = event => {
	return async (dispatch, getState) => {
		const { name, uid } = getState().auth;

		try {
			const response = await fetchWithToken('events', event, 'POST');
			const body = await response.json();

			if (body.ok) {
				dispatch(
					addNewEvent({
						...event,
						id: body.event.id,
						user: {
							_id: uid,
							name,
						},
					})
				);
			} else {
				toast.error(body.msg);
			}
		} catch (error) {
			console.log(error);
			toast.error(error);
		}
	};
};

const addNewEvent = event => ({
	type: types.eventAddNew,
	payload: event,
});

export const setActiveEvent = event => ({
	type: types.eventSetActive,
	payload: event,
});

export const cleanActiveEvent = () => ({
	type: types.eventCleanActive,
});

/*
    UPDATE EVENT
*/
export const eventStartUpdate = event => {
	return async dispatch => {
		try {
			const response = await fetchWithToken(
				`events/${event.id}`,
				event,
				'PUT'
			);

			const body = await response.json();
			if (body.ok) {
				dispatch(updatedEvent(event));
			} else {
				toast.error(body.msg);
			}
		} catch (error) {
			console.log(error);
			toast.error(error);
		}
	};
};

const updatedEvent = event => ({
	type: types.eventUpdated,
	payload: event,
});

/*
    DELETE EVENT
*/

export const eventStartDelete = () => {
	return async (dispatch, getState) => {
		const { activeEvent } = getState().calendar;
		try {
			const response = await fetchWithToken(
				`events/${activeEvent.id}`,
				{},
				'DELETE'
			);

			const { ok, msg } = await response.json();

			if (ok) {
				dispatch(deletedEvent());
			} else {
				toast.error(msg);
			}
		} catch (error) {
			console.log(error);
			toast.error(error);
		}
	};
};

const deletedEvent = () => ({
	type: types.eventDeleted,
});

export const eventsCleanAll = () => ({
	type: types.eventCleanAll,
});

/*
    LOADING THE EVENTS FROM DB
*/

export const eventStartLoading = () => {
	return async dispatch => {
		try {
			const response = await fetchWithToken('events');

			const body = await response.json();

			if (body.ok) {
				const eventsPrepared = prepareEvents(body.events);
				dispatch(eventsLoaded(eventsPrepared));
			} else {
				toast.error(body.msg);
			}
		} catch (error) {
			console.log(error);
		}
	};
};

const eventsLoaded = events => ({
	type: types.eventLoaded,
	payload: events,
});
