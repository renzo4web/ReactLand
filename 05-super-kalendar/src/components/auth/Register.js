import React from "react";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";

const Register = ({ onFinish }) => {
    return (
        <Form className="bg-white p-3 rounded text-dark">
            <h2>Register</h2>

            <Form.Group className="mb-3" controlId="formBasicEmail">
                <Form.Label>Email address</Form.Label>
                <Form.Control type="email" placeholder="Enter email" />
                <Form.Text className="text-muted">
                    Use a temporary email if necessary
                </Form.Text>
            </Form.Group>

            <Form.Group className="mb-3" controlId="formBasicPassword">
                <Form.Label>Password</Form.Label>
                <Form.Control type="password" placeholder="Password" />
            </Form.Group>

            <Form.Group className="mb-3" controlId="formBasicPassword">
                <Form.Label>Confirm Password</Form.Label>
                <Form.Control type="password" placeholder="Password" />
            </Form.Group>

            <Button variant="primary" type="submit">
                Submit
            </Button>
        </Form>
    );
};

export default Register;
