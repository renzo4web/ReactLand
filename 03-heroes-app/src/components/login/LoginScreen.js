import {
  Button,
  Heading,
  Center,
  Flex,
  Image,
  Box,
  Input,
  Text,
} from '@chakra-ui/react';
import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { useAuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';
import img from './../../assets/heroes/dc-arrow.jpg';

const LoginScreen = () => {
  const [, dispatch] = useAuthContext();
  const [name, setName] = useState('');
  const history = useHistory();
  const lastPath =
    JSON.parse(localStorage.getItem('lastPath')) || '/';

  const handleLogin = () => {
    dispatch({
      type: types.login,
      payload: {
        name,
      },
    });
    history.replace(lastPath);
  };

  return (
    <Center h="100vh">
      <Box>
        <Flex direction="column">
          <Heading mb="20px" as="h1">
            Login
          </Heading>
          <Image
            my="1em"
            boxSize="200px"
            objectFit="cover"
            src={img}
            alt="Arrow"
          />
          <Text>Name:</Text>
          <Input
            size="md"
            placeholder="Mr Marvel"
            value={name}
            onChange={({ target }) => setName(target.value)}
            onBlur={({ target }) => setName(target.value)}
            mb="1em"
          />
          <Button
            onClick={() => handleLogin()}
            colorScheme="teal"
          >
            Login
          </Button>
        </Flex>
      </Box>
    </Center>
  );
};

export default LoginScreen;
