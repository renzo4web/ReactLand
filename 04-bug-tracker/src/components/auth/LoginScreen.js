import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
import { removeError, setError } from '../../actions/ui';

const LoginScreen = () => {
    const initialForm = {
        email: '',
        password: '',
    };
    const { msgError, loading } = useSelector((state) => state.ui);
    const dispatch = useDispatch();
    const [formValues, handleInputChange] = useForm(initialForm);

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startLoginEmailPassword(email, password));
        }
    };

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    };

    const isFormValid = () => {
        if (!validator.isEmail(email)) {
            dispatch(setError('Email is not valid'));
            return false;
        }

        if (password.length < 5) {
            dispatch(
                setError(
                    'Password should be at least 5 characters long and match each other'
                )
            );
            return false;
        }

        dispatch(removeError());
        return true;
    };

    return (
        <>
            <h1>Login</h1>

            {msgError && (
                <strong className='ui__alert-input'>{msgError}</strong>
            )}
            <form onSubmit={handleLogin}>
                <div className='ui__input-group'>
                    <label htmlFor='email'>Email</label>
                    <input
                        value={email}
                        placeholder='test@gmail.com'
                        onChange={handleInputChange}
                        name='email'
                        id='email'
                        type='text'
                    />
                </div>

                <div className='ui__input-group'>
                    <label htmlFor='password'>Password</label>
                    <input
                        value={password}
                        onChange={handleInputChange}
                        name='password'
                        id='password'
                        type='password'
                    />
                </div>

                <button disabled={loading} type='submit' className='auth__btn'>
                    Login
                </button>

                <hr />

                <div>
                    <p>Login with social networks</p>

                    <div onClick={handleGoogleLogin} className='google-btn'>
                        <div className='google-icon-wrapper'>
                            <img
                                className='google-icon'
                                src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
                                alt='google button'
                            />
                        </div>
                        <p className='btn-text'>
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to='/auth/register'>Create new account</Link>
            </form>
        </>
    );
};

export default LoginScreen;
