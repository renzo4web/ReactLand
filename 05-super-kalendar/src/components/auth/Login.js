import React from 'react';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import { useForm } from '../../hooks/useForm';
import { startLogin } from '../../actions/auth';
import { useDispatch } from 'react-redux';

const Login = ({ onFinish }) => {
	const dispatch = useDispatch();
	const [formLoginVals, handleLoginInputChange] = useForm({
		lEmail: 'test@gmail.com',
		lPassword: '123456',
	});

	const { lEmail, lPassword } = formLoginVals;

	const handleLogin = async e => {
		e.preventDefault();
		dispatch(startLogin(lEmail, lPassword));
	};

	return (
		<Form
			onSubmit={handleLogin}
			className="bg-dark p-3 rounded text-white flex-fill"
		>
			<h2>Login</h2>
			<Form.Group className="mb-3" controlId="formBasicEmail">
				<Form.Label>Email address</Form.Label>
				<Form.Control
					type="email"
					placeholder="Enter email"
					name="lEmail"
					value={lEmail}
					onChange={handleLoginInputChange}
					onBlur={handleLoginInputChange}
				/>
				<Form.Text className="text-muted">
					We'll never share your email with anyone else.
				</Form.Text>
			</Form.Group>

			<Form.Group className="mb-3" controlId="formBasicPassword">
				<Form.Label>Password</Form.Label>
				<Form.Control
					type="password"
					placeholder="Password"
					name="lPassword"
					value={lPassword}
					onChange={handleLoginInputChange}
					onBlur={handleLoginInputChange}
				/>
			</Form.Group>
			<Button variant="primary" type="submit">
				Submit
			</Button>
		</Form>
	);
};

export default Login;
