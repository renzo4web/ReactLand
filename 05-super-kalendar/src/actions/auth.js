import { types } from '../types/types';
import { fetchWithoutToken, fetchWithToken } from '../helpers/fetch';
import toast from 'react-hot-toast';
import { eventsCleanAll } from './events';

export const startLogin = (email, password) => {
	return async dispatch => {
		// http://localhost:4000/api/auth
		const response = await fetchWithoutToken(
			'auth',
			{ email, password },
			'POST'
		);
		const body = await response.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			// keep track initial date of the token created
			localStorage.setItem('token-init-date', new Date().getTime());

			const { uid, name } = body;

			dispatch(
				login({
					uid,
					name,
				})
			);
		} else {
			toast.error(body.msg);
		}
	};
};

export const startRegister = (name, email, password) => {
	return async dispatch => {
		const response = await fetchWithoutToken(
			'auth/new',
			{ email, password, name },
			'POST'
		);

		const body = await response.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			// keep track initial date of the token created
			localStorage.setItem('token-init-date', new Date().getTime());

			const { uid, name } = body;

			dispatch(
				login({
					uid,
					name,
				})
			);
		} else {
			toast.error(body.msg);
		}
	};
};

export const startChecking = () => {
	return async dispatch => {
		const response = await fetchWithToken('auth/renew');
		const body = await response.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			// keep track initial date of the token created
			localStorage.setItem('token-init-date', new Date().getTime());

			const { uid, name } = body;

			dispatch(
				login({
					uid,
					name,
				})
			);
		} else {
			dispatch(checkingFinish());
		}
	};
};

const checkingFinish = () => ({
	type: types.authCheckingFinish,
});

export const startLogout = () => {
	return dispatch => {
		localStorage.clear();
		dispatch(logout());
		dispatch(eventsCleanAll());
	};
};

const logout = () => ({
	type: types.authLogout,
});

const login = user => ({
	type: types.authLogin,
	payload: user,
});
