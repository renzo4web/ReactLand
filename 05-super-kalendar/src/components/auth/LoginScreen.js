import React from 'react';
import Login from './Login';
import Register from './Register';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

const LoginScreen = () => {
	return (
		<Container className="my-auto">
			<Row
				className="my-auto bg-primary text-white p-5 rounded"
				as="main"
			>
				<Col sm>
					<Login />
				</Col>
				<Col sm>
					<Register />
				</Col>
			</Row>
		</Container>
	);
};

export default LoginScreen;
