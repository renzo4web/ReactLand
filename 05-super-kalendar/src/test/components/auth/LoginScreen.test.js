import { mount } from 'enzyme';
import React from 'react';
import toast from 'react-hot-toast';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { startLogin, startRegister } from '../../../actions/auth';
import LoginScreen from '../../../components/auth/LoginScreen';

jest.mock('../../../actions/auth', () => ({
	startLogin: jest.fn(),
	startRegister: jest.fn(),
}));

jest.mock('react-hot-toast', () => ({
	error: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
	<Provider store={store}>
		<LoginScreen />
	</Provider>
);

describe('Test components/auth/LoginScreen', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	test('should render properly', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('should dispatch starLogin', () => {
		wrapper.find('form.bg-dark').simulate('submit', {
			preventDefault() {},
		});
		expect(startLogin).toHaveBeenCalled();
	});

	test('should not register if passwords not match', () => {
		wrapper.find('form.bg-white input[name="rName"]').simulate('change', {
			target: {
				name: 'rName',
				value: 'Test',
			},
		});

		wrapper.find('form.bg-white input[name="rEmail"]').simulate('change', {
			target: {
				name: 'rEmail',
				value: 'test@gmail.com',
			},
		});
		wrapper
			.find('form.bg-white input[name="rPassword1"]')
			.simulate('change', {
				target: {
					name: 'rPassword1',
					value: '11111',
				},
			});
		wrapper
			.find('form.bg-white input[name="rPassword2"]')
			.simulate('change', {
				target: {
					name: 'rPassword2',
					value: 'ZZZZ',
				},
			});

		// SUBMIT
		wrapper.find('form.bg-white').simulate('submit', {
			preventDefault() {},
		});

		expect(startRegister).not.toHaveBeenCalled();
		expect(toast.error).toHaveBeenCalled();
	});

	test('should register the user when the passwords match', () => {
		wrapper.find('form.bg-white input[name="rName"]').simulate('change', {
			target: {
				name: 'rName',
				value: 'Test',
			},
		});

		wrapper.find('form.bg-white input[name="rEmail"]').simulate('change', {
			target: {
				name: 'rEmail',
				value: 'test@gmail.com',
			},
		});
		wrapper
			.find('form.bg-white input[name="rPassword1"]')
			.simulate('change', {
				target: {
					name: 'rPassword1',
					value: '11111',
				},
			});
		wrapper
			.find('form.bg-white input[name="rPassword2"]')
			.simulate('change', {
				target: {
					name: 'rPassword2',
					value: '11111',
				},
			});

		// SUBMIT
		wrapper.find('form.bg-white').simulate('submit', {
			preventDefault() {},
		});

		expect(startRegister).toHaveBeenCalled();
		expect(toast.error).not.toHaveBeenCalled();
	});
});
