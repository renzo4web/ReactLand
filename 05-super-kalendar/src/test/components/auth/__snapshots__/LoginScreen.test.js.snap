// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test components/auth/LoginScreen should render properly 1`] = `
<div
  className="my-auto container"
>
  <main
    className="my-auto bg-primary text-white p-5 rounded row"
  >
    <div
      className="col-sm"
    >
      <form
        className="bg-dark p-3 rounded text-white flex-fill"
        onSubmit={[Function]}
      >
        <h2>
          Login
        </h2>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formBasicEmail"
          >
            Email address
          </label>
          <input
            className="form-control"
            id="formBasicEmail"
            name="lEmail"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Enter email"
            type="email"
            value="test@gmail.com"
          />
          <small
            className="text-muted form-text"
          >
            We'll never share your email with anyone else.
          </small>
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formBasicPassword"
          >
            Password
          </label>
          <input
            className="form-control"
            id="formBasicPassword"
            name="lPassword"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value="123456"
          />
        </div>
        <button
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
    <div
      className="col-sm"
    >
      <form
        className="bg-white p-3 rounded text-dark"
        onSubmit={[Function]}
      >
        <h2>
          Register
        </h2>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formBasicText"
          >
            Name
          </label>
          <input
            className="form-control"
            id="formBasicText"
            name="rName"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Enter your name"
            type="text"
            value=""
          />
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formBasicEmail"
          >
            Email address
          </label>
          <input
            className="form-control"
            id="formBasicEmail"
            name="rEmail"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Enter email"
            type="email"
            value=""
          />
          <small
            className="text-muted form-text"
          >
            Use a temporary email if necessary
          </small>
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formBasicPassword"
          >
            Password
          </label>
          <input
            className="form-control"
            id="formBasicPassword"
            name="rPassword1"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formBasicPassword"
          >
            Confirm Password
          </label>
          <input
            className="form-control"
            id="formBasicPassword"
            name="rPassword2"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <button
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </main>
</div>
`;
